core/modules/map.rs

```diff
@@ -702,155 +700,115 @@ impl ModuleMap {
     );
 
     let (sender, receiver) = oneshot::channel();
+    let receiver = receiver.map(|res| {
+      res.unwrap_or_else(|_| {
+        bail!("Cannot evaluate module, because JavaScript execution has been terminated")
+      }
+    )});
 
-    // IMPORTANT: Top-level-await is enabled, which means that return value
-    // of module evaluation is a promise.
-    //
-    // Because that promise is created internally by V8, when error occurs during
-    // module evaluation the promise is rejected, and since the promise has no rejection
-    // handler it will result in call to `bindings::promise_reject_callback` adding
-    // the promise to pending promise rejection table - meaning JsRuntime will return
-    // error on next poll().
-    //
-    // This situation is not desirable as we want to manually return error at the
-    // end of this function to handle it further. It means we need to manually
-    // remove this promise from pending promise rejection table.
-    //
-    // For more details see:
-    // https://github.com/denoland/deno/issues/4908
-    // https://v8.dev/features/top-level-await#module-execution-order
-    {
-      let mut mod_evaluate = self.pending_mod_evaluate.borrow_mut();
-      assert!(
-        mod_evaluate.is_none(),
-        "There is already pending top level module evaluation"
-      );
-      *mod_evaluate = Some(ModEvaluate {
-        promise: None,
-        has_evaluated: false,
-        handled_promise_rejections: vec![],
-        sender,
-      });
-    }
-
-    let maybe_value = module.evaluate(tc_scope);
+    let Some(value) = module.evaluate(tc_scope) else {
+      if tc_scope.has_terminated() || tc_scope.is_execution_terminating() {
+        let undefined = v8::undefined(tc_scope).into();
+        _ = sender
+          .send(exception_to_err_result(tc_scope, undefined, true, false));
+      } else {
+        debug_assert_eq!(module.get_status(), v8::ModuleStatus::Errored);
```

Comment:
Why `debug_assert_eq` and not `assert_eq`?

Reply:
I want to avoid too many asserts in release mode -- we assert quite a bit and our debug tests should be sufficient to catch these. Generally we should avoid `assert` because it bloats the codebase and inhibits the optimizer.